### PR TITLE
WT-11398 Application thread stuck doing eviction because snapshot is old

### PR DESCRIPTION
**Pull request by raviprakashgiri29**
This pull request has two sets of changes.

- Save the application threads snapshots when application threads enter into eviction flow to acquire a new snapshot and then switch back to the original snapshots after application threads come out of eviction flow.
- Update the existing snap_max, snap_min in the txn structure to use the new WT_TXN_SNAPSHOT_DATA structure.

### CODE REVIEW

File: src/txn/txn.c [hunk 1]
```diff
@@ -308,6 +309,21 @@ __wt_txn_get_snapshot(WT_SESSION_IMPL *session)
     __txn_get_snapshot_int(session, true);
 }
 
+/*
+ * __wt_swap_snapshot --
+ *     Swap the snapshot pointers.
+ */
+void
+__wt_swap_snapshot(uint64_t **snap_a, uint64_t **snap_b)
```
Comment (quchenhao) [inline src/txn/txn.c:317]: Does this need to be with __wt prefix?

Comment (raviprakashgiri29) [reply src/txn/txn.c]: Nice catch, thanks @quchenhao. Fixed in 074dc0fde461582847671b3de4080b87a9f331f7. Changed the function name to `__txn_swap_snapshot`

File: src/txn/txn.c [hunk 2]
```diff
@@ -308,6 +309,21 @@ __wt_txn_get_snapshot(WT_SESSION_IMPL *session)
     __txn_get_snapshot_int(session, true);
 }
 
+/*
+ * __txn_swap_snapshot --
+ *     Swap the snapshot pointers.
+ */
+static void
+__txn_swap_snapshot(uint64_t **snap_a, uint64_t **snap_b)
```
Comment (quchenhao) [inline src/txn/txn.c:317]: May be declare this as a incline function and move to txn_inline

Comment (raviprakashgiri29) [reply src/txn/txn.c]: Moved __txn_swap_snapshot to inline function in txn_inline in 87d8ff44e854eafef0bf73911ada633b4d983628 commit

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/include/txn.h [hunk 3]
```diff
@@ -242,6 +242,22 @@ struct __wt_txn_op {
     uint32_t flags;
 };
 
+/*
+ * WT_TXN_SNAPSHOT_DATA --
+ *	A structure to store the transactions snapshot details.
+ */
+struct __wt_txn_snapshot_data {
```
Comment (bvpvamsikrishna) [inline src/include/txn.h:249]: my suggestion is to name the struct as __wt_txn_snapshot, just remove _data at the end.

File: src/include/txn.h [hunk 4]
```diff
@@ -255,17 +271,14 @@ struct __wt_txn {
...
-     *	ids < snap_min are visible,
-     *	everything else is visible unless it is in the snapshot.
-     */
-    uint64_t snap_min, snap_max;
-    uint64_t *snapshot;
-    uint32_t snapshot_count;
     uint32_t txn_logsync; /* Log sync configuration */
 
+    /* Snapshot data. */
+    WT_TXN_SNAPSHOT_DATA snapshot_data;
```
Comment (bvpvamsikrishna) [inline src/include/txn.h:277]: suggest to remove _data in this as well as next.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm.
Please change the name to remove _data at the end of new struct and new variable.s

File: src/include/txn.h (hunk 3 above)
Comment (raviprakashgiri29) [reply src/include/txn.h]: Fixed in 3523185a21d3fff8e71151f767b27376ee91d929

File: src/include/txn.h (hunk 4 above)
Comment (raviprakashgiri29) [reply src/include/txn.h]: Fixed in 3523185a21d3fff8e71151f767b27376ee91d929

**Pull request merged**